nemo/collections/nlp/models/information_retrieval/bert_embedding_model.py

```diff
@@ -0,0 +1,188 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+
+from nemo.collections.nlp.models.language_modeling.megatron.bert.bert_model import (
+    BertModel,
+    MCoreBertModelWrapper,
+    bert_extended_attention_mask,
+)
+from nemo.collections.nlp.modules.common.megatron.utils import ApexGuardDefaults, build_position_ids
+
+try:
+    from megatron.core import ModelParallelConfig, parallel_state
+
+    HAVE_MEGATRON_CORE = True
+
+except (ImportError, ModuleNotFoundError):
+
+    ModelParallelConfig = ApexGuardDefaults
+
+    HAVE_MEGATRON_CORE = False
+
+
+class BertEmbeddingHead(nn.Module):
+    """Performs mean pooling on the token embeddings.
+    """
+
+    def __init__(
+        self, word_embedding_dimension: int, pooling_mode_mean_tokens: bool = True,
+    ):
+        super(BertEmbeddingHead, self).__init__()
+
+        self.config_keys = [
+            "word_embedding_dimension",
+            "pooling_mode_mean_tokens",
+        ]
+        self.word_embedding_dimension = word_embedding_dimension
+        self.pooling_mode_mean_tokens = pooling_mode_mean_tokens
+
+    def forward(self, token_embeddings: Tensor, attention_mask: Tensor):
+
+        token_embeddings = token_embeddings.permute(1, 0, 2)
+        input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
+        sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
+
+        sum_mask = input_mask_expanded.sum(1)
+
+        sum_mask = torch.clamp(sum_mask, min=1e-9)
+
+        output_vector = sum_embeddings / sum_mask
+
+        output_vector = F.normalize(output_vector, p=2, dim=1)
+
+        return output_vector
+
+    def __repr__(self):
+        return "Pooling({}) and Normalize".format(self.get_config_dict())
+
+    def get_config_dict(self):
+        return {key: self.__dict__[key] for key in self.config_keys}
+
+
+class MCoreBertEmbeddingModel(MCoreBertModelWrapper):
+    def __init__(self, *args, **kwargs):
+
+        super(MCoreBertEmbeddingModel, self).__init__(*args, **kwargs)
+        self.post_process = False
+        self.binary_head = None
+        self.lm_head = None
+        self.output_layer = None
+        self.encoder.final_layernorm = None
+        self.encoder.post_process = False
+        self.embedding_head = BertEmbeddingHead(
+            word_embedding_dimension=self.config.hidden_size, pooling_mode_mean_tokens=True,
+        )
+
+    def forward(
+        self,
+        input_ids: Tensor,
+        attention_mask: Tensor,
+        tokentype_ids: Tensor = None,
+        lm_labels: Tensor = None,
+        inference_params=None,
+    ):
+        """Forward function of BERT model
+
```

Comment:
Can you use base class forward function() . It does the same things, you can just set the post_prcoess flag to false, and then pass the output through the embedding head.